File: 1. 단계별로_풀어보기/30~39/31. 동적 계획법 2/6. 7579.js
// 입력
const fs = require('fs')
const inputArr = fs.readFileSync(0).toString().trim().split('\n')

// 입력 파싱
const [N, M] = inputArr[0].trim().split(' ').map(Number)
const A = inputArr[1].trim().split(' ').map(Number)
const C = inputArr[2].trim().split(' ').map(Number)

// 코스트마다 최대 메모리 저장할 배열
const resultArr = Array.from({length: N + 1}, () => Array.from({length: 10001}).fill(0))

for (let i = 0; i < N; i++) {
  const mem = A[i]
  const cost = C[i]

  resultArr[i + 1][cost] = mem

  for (let c = 0; c + cost <= 10000; c++) {
    if (resultArr[i][c] > 0) {
      resultArr[i + 1][c] = Math.max(resultArr[i + 1][c], resultArr[i][c])
      resultArr[i + 1][c + cost] = Math.max(resultArr[i + 1][c + cost], resultArr[i][c] + mem)
    }
  }
}

for (let cost = 0; cost <= 10000; cost++) {
  if (resultArr[N][cost] >= M) {
    process.stdout.write(`${cost}`)
    break
  }
}

File: 1. 단계별로_풀어보기/30~39/31. 동적 계획법 2/6. 7579.py
import sys

sys.setrecursionlimit(10000)

input = sys.stdin.readline
output = sys.stdout.write

N, M = map(int, input().split())
A = list(map(int, input().split()))
C = list(map(int, input().split()))

resultArr = list(list(None for i in range(10001)) for j in range(N + 1))

resultArr[0][0] = 0

def getResult(idx, c):
  if resultArr[idx][c] != None:
    return resultArr[idx][c]

  if idx == 0:
    return -1

  cost = C[idx - 1]
  mem = A[idx - 1]

  used = getResult(idx - 1, c - cost) if c >= cost else -1
  unUsed = getResult(idx - 1, c)

  if used > -1:
    used += mem

  resultArr[idx][c] = max(used, unUsed)
  return resultArr[idx][c]
  
output(f"{min(cost for cost in range(10001) if getResult(N, cost) >= M)}")
